### PR TITLE
Make WaitForIdle() methods virtual so a user can extend it's functionality

### PR DESCRIPTION
**Pull request by StefanOssendorf**
* WaitForIdle() methods are now virtual
* Reorganze WaitForIdle() methods to reduce code duplication

@rockfordlhotka 
During the changes I've become unhappy with the currenty `WaitForIdle()` structure, because a user would now have to override two methods to get consistent behavior. The methods with `TimeSpan` and `CancellationToken` because the first throws a `TimeoutException` and the latter a `TaskCanceledException`. I understand _why_ both implementation use different exceptions but from a user perspective it's awkward to override two methods to have consistent behavior. Any ideas?

### CODE REVIEW

File: Source/Csla/Core/BusinessBase.cs [hunk 1]
```diff
@@ -1227,6 +1227,33 @@ public virtual BrokenRulesCollection BrokenRulesCollection
...
+    {
+      var cslaOptions = ApplicationContext.GetRequiredService<Configuration.CslaOptions>();
+      await WaitForIdle(TimeSpan.FromSeconds(cslaOptions.DefaultWaitForIdleTimeoutInSeconds)).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Await this method to ensure business object is not busy.
+    /// </summary>
+    /// <param name="timeout">Timeout duration</param>
+    public virtual Task WaitForIdle(TimeSpan timeout)
```
Comment (rockfordlhotka) [inline Source/Csla/Core/BusinessBase.cs:1243]: Is there a way to have just one virtual overload?

I've done this before, with two virtual overloads that "do the same thing" and it becomes a serious point of confusion and bugs for people who overload the method(s) over time.

Comment (StefanOssendorf) [reply Source/Csla/Core/BusinessBase.cs:1243]: Yeah. I'll think about a proper solution :)

File: Source/Csla/Core/BusinessBase.cs [hunk 2]
```diff
@@ -1240,9 +1240,9 @@ public async Task WaitForIdle()
     /// Await this method to ensure business object is not busy.
     /// </summary>
     /// <param name="timeout">Timeout duration</param>
-    public virtual Task WaitForIdle(TimeSpan timeout)
+    public Task WaitForIdle(TimeSpan timeout)
```
Comment (rockfordlhotka) [inline Source/Csla/Core/BusinessBase.cs:1243]: If this overload doesn't invoke the virtual overload, then the user can't replace this implementation? Only the cancelation token overload?

Can't both public overloads maybe invoke a `protected virtual` that someone can override?

**Review by rockfordlhotka — CHANGES_REQUESTED** [review]

File: Source/Csla/Core/BusinessBase.cs (hunk 2 above)
Comment (StefanOssendorf) [reply Source/Csla/Core/BusinessBase.cs:1243]: Huh? The first Parameter is using the `WaitForIdle(CancellationToken)` method so we have only one method to override.

Comment (rockfordlhotka) [reply Source/Csla/Core/BusinessBase.cs:1247]: Line 1245 confused me. I got it now.

**Review by rockfordlhotka — APPROVED** [review]

**Pull request merged**